src/connections/connections.service.ts

```diff
@@ -0,0 +1,177 @@
+import {
+  BadRequestException,
+  Injectable,
+  NotFoundException,
+} from '@nestjs/common';
+import { AuthType, User } from '@prisma/client';
+import { PrismaService } from 'src/prisma/prisma.service';
+import { ConnectionDto } from './DTO/Connection.dto';
+import { ApiTags } from '@nestjs/swagger';
+import { ProfileFetcherDelegator } from 'src/user/profile-fetcher/delegator.profile-fetcher';
+import { v4 } from 'uuid';
+
+@Injectable()
+@ApiTags('Connections controller')
+export class ConnectionsService {
+  constructor(private readonly prisma: PrismaService) {}
+
+  // retuns an object to be used with prisma's include
+  private includeWithUserConnection(currentUserId?: User['id']) {
+    return {
+      _count: {
+        select: {
+          followings: true,
+          reviewsReceived: true,
+        },
+      },
+      followers: currentUserId
+        ? {
+            where: {
+              followerId: currentUserId,
+            },
+          }
+        : false,
+    };
+  }
+
+  // transforms an user from db to a Connection DTO
+  private transformUserConnection(
+    user: User & { _count: { followings: number; reviewsReceived: number } } & {
+      followers?: { id: number; followerId: string; followingId: string }[];
+    },
+  ): ConnectionDto {
+    return {
+      id: user.id,
+      email: user.email,
+      name: user.name,
+      headline: user.headline,
+      profilePictureUrl: user.profilePictureUrl,
+      isEmailVerified: user.isEmailVerified,
+      joinedAt: user.joinedAt,
+      authType: user.authType,
+      reviewsCount: user._count.reviewsReceived,
+      connectionsCount: user._count.followings,
+      isConnection: user.followers && user.followers.length !== 0,
+    };
+  }
+
+  async getConnection(connectionId: User['id'], currentUserId: User['id']) {
+    const user = await this.prisma.user.findUnique({
+      where: {
+        id: connectionId,
+      },
+      include: this.includeWithUserConnection(currentUserId),
+    });
+
+    if (!user) {
+      throw new NotFoundException(`User with ${connectionId} not found`);
+    }
+
+    return this.transformUserConnection(user);
+  }
+
+  async getUserConnections(userId: User['id']) {
+    const connections = await this.prisma.connection.findMany({
+      where: {
+        followerId: userId,
+      },
+      include: {
+        following: {
+          include: this.includeWithUserConnection(userId),
+        },
+      },
+    });
+
+    return connections.map((c) => this.transformUserConnection(c.following));
+  }
+
+  async connectWithUser(currentUserId: User['id'], userId: User['id']) {
+    await this.prisma.connection.upsert({
+      where: {
+        followerId_followingId: {
+          followerId: currentUserId,
+          followingId: userId,
+        },
+      },
+      update: {},
+      create: {
+        followerId: currentUserId,
+        followingId: userId,
+      },
+    });
+
+    return this.getConnection(userId, currentUserId);
```

Comment:
This will try to query the db twice. Instead, you can do [this](https://www.prisma.io/docs/orm/prisma-client/queries/crud#update-or-create-records)